### PR TITLE
feat(api): New AddToken endpoint

### PR DESCRIPTION
**Pull request by a-launois**
## Description

New API endpoint AddToken
```
POST /administration/tokens
{
  "name":" token name ",
  "user_id": 20,
  "expiration_date": "2024-09-10T00:00:00+00:00"
}
```

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    {
+        $this->validation->assertIsValidName($request->name);
+        $this->validation->assertIsValidUser($request->userId);
+
+        $localProviderId = $this->providerFactory->create(Provider::LOCAL)->getConfiguration()->getId();
+        $tokenString = Encryption::generateRandomString();
+        $creationDate = new \DateTimeImmutable();
+        $expirationDate = new \DateTimeImmutable($request->expirationDate->format('Y-m-d'));
+
+        $this->validation->assertIsValidExpirationDate($expirationDate, $creationDate);
```
Comment (callapa) [inline centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php:119]: Must be defined in the NewToken class

File: centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+     * @throws \Throwable
+     *
+     * @return string
+     */
+    private function createToken(AddTokenRequest $request): string
+    {
+        $this->validation->assertIsValidName($request->name);
+        $this->validation->assertIsValidUser($request->userId);
+
+        $localProviderId = $this->providerFactory->create(Provider::LOCAL)->getConfiguration()->getId();
```
Comment (callapa) [inline centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php:114]: No longer needed

File: centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+     * @throws \Throwable
+     *
+     * @return AddTokenResponse
+     */
+    private function createResponse(string $tokenString): AddTokenResponse
+    {
+        if (! ($apiToken = $this->readTokenRepository->find($tokenString))) {
+            throw TokenException::errorWhileRetrievingObject();
+        }
+        $userNames = $this->readContactRepository->findNamesByIds($apiToken->getUserId());
```
Comment (callapa) [inline centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php:151]: UserName can be retrieve directly from the repository

File: centreon/src/Core/Security/Token/Domain/Model/NewToken.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+     */
+    public function __construct(
+        private string $token,
+        private \DateTimeImmutable $creationDate,
+        private \DateTimeImmutable $expirationDate,
+        private int $userId,
+        private int $configurationProviderId,
+        private string $name,
+        private int $creatorId,
+        private string $creatorName
```
Comment (abuathier) [inline centreon/src/Core/Security/Token/Domain/Model/NewToken.php:51]: Immutable by default is possible please.
👉🏻  `readonly`

File: centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php (hunk 2 above)
Comment (a-launois) [reply centreon/src/Core/Security/Token/Application/UseCase/AddToken/AddToken.php]: Has been removed from Token model, still used in the newToken model

File: centreon/src/Core/Security/Token/Infrastructure/Repository/DbReadTokenRepository.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+     */
+    public function existsByName(string $tokenName): bool
+    {
+        $statement = $this->db->prepare($this->translateDbName(
+            <<<'SQL'
+                SELECT 1
+                FROM `:db`.security_authentication_tokens sat
+                WHERE sat.token_name = :tokenName
+                    AND sat.token_type = 'manual'
+                SQL
```
Comment (callapa) [inline centreon/src/Core/Security/Token/Infrastructure/Repository/DbReadTokenRepository.php:113]: Uniqueness should also depend on user Id

**Review by callapa — APPROVED** [review]

**Review by abuathier — APPROVED** [review]

**Review by callapa — DISMISSED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 6]
```diff
@@ -18128,4 +18128,22 @@ msgid "View by host"
...
+msgstr "Erreur lors de l'ajout d'un token"
+
+msgid "You are not allowed to add tokens"
+msgstr "Vous n'êtes pas autorisé à ajouter des tokens"
+
+msgid "Error while retrieving a token"
+msgstr "Erreur lors de la récupération du token"
+
+msgid "The token name '%s' already exists"
+msgstr "Le nom de token '%s' exist déjà"
```
Comment (cg-tw) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18143]:
```suggestion
msgstr "Le nom de token '%s' existe déjà"
```

File: centreon/tests/php/Core/Security/Token/Application/UseCase/AddToken/AddTokenTest.php [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+
+    ($this->useCase)($this->request, $this->presenter);
+
+    expect($this->presenter->response)
+        ->toBeInstanceOf(ConflictResponse::class)
+        ->and($this->presenter->response->getMessage())
+        ->toBe(TokenException::invalidUserId($this->request->userId)->getMessage());
+});
+
+it('should present a ConflictResponse when a creator can not manage user\'s tokens', function (): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Security/Token/Application/UseCase/AddToken/AddTokenTest.php:185]:
```suggestion
it('should present a ConflictResponse when a creator cannot manage user\'s tokens', function (): void {
```

**Review by cg-tw — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**